framework/include/utils/FactorizedSymmetricRankTwoTensor.h

```diff
@@ -0,0 +1,300 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "RankTwoTensor.h"
+
+// forward declarations
+template <typename>
+class FactorizedSymmetricRankTwoTensorTempl;
+
+namespace MathUtils
+{
+/// natural log of a factorized RankTwoTensor
+template <typename T>
+FactorizedSymmetricRankTwoTensorTempl<T> log(const FactorizedSymmetricRankTwoTensorTempl<T> &);
+
+/// exponentiated a factorized RankTwoTensor
+template <typename T>
+FactorizedSymmetricRankTwoTensorTempl<T> exp(const FactorizedSymmetricRankTwoTensorTempl<T> &);
+
+/// a factorized RankTwoTensor raised to a power
+template <typename T, typename T2, typename std::enable_if<ScalarTraits<T2>::value, int>::type = 0>
+FactorizedSymmetricRankTwoTensorTempl<typename CompareTypes<T, T2>::supertype>
+pow(const FactorizedSymmetricRankTwoTensorTempl<T> &, const T2 & p);
+
+/// sqrt of a factorized RankTwoTensor
+template <typename T>
+FactorizedSymmetricRankTwoTensorTempl<T> sqrt(const FactorizedSymmetricRankTwoTensorTempl<T> &);
+
+/// cbrt of a factorized RankTwoTensor
+template <typename T>
+FactorizedSymmetricRankTwoTensorTempl<T> cbrt(const FactorizedSymmetricRankTwoTensorTempl<T> &);
+} // end namespace MathUtils
+
+/**
+ * FactorizedSymmetricRankTwoTensorTempl is designed to perform the spectral decomposition of an
+ * underlying RankTwoTensorTempl and reuse its bases for future operations if possible.
+ *
+ * Only operations that reuses the known factorization are provided. Otherwise, you will need to
+ * first retrieve the underlying RankTwoTensorTempl to perform the operation.
+ */
+template <typename T>
+class FactorizedSymmetricRankTwoTensorTempl
+{
+public:
+  /// No default constructor
+  FactorizedSymmetricRankTwoTensorTempl() = delete;
+
+  /// Copy constructor
+  FactorizedSymmetricRankTwoTensorTempl(const FactorizedSymmetricRankTwoTensorTempl<T> & A) =
+      default;
+
+  /// Construct from RankTwoTensorTempl<T> if the factorization isn't known a priori
+  FactorizedSymmetricRankTwoTensorTempl(const RankTwoTensorTempl<T> & A)
+  {
+#ifdef DEBUG
+    RankTwoTensorTempl<T> error = A - A.transpose();
+    if (!MooseUtils::absoluteFuzzyEqual(error.norm(), 0))
+      mooseError("The tensor is not symmetric.");
+#endif
+    A.symmetricEigenvaluesEigenvectors(_eigvals, _eigvecs);
+  }
+
+  /// Construct from RankTwoTensorTempl<T> if the factorization is known
+  FactorizedSymmetricRankTwoTensorTempl(const std::vector<T> & eigvals,
+                                        const RankTwoTensorTempl<T> & eigvecs)
+    : _eigvals(eigvals), _eigvecs(eigvecs)
+  {
+  }
+
+  // @{ getters
+  RankTwoTensorTempl<T> get() const { return assemble(); }
```

Comment:
I think it would be forward looking to template on the tensor type itself (that could be a template template argument, and then another argument for the scalar type. Or better yet, we add a `value_type` member to `RankTwoTensorTempl<T>` that resolves to `T`. 

That way we can use the wrapper if that SymmetricRankTwoTensor PR ever gets merged. In that context I'd suggest calling the instantiation of you class `FactorizedRankTwoTensor` to distinguish it from a future `FactorizedSymmetricRankTwoTensor`, that is based on `SymmetricRankTwoTensor`.

Reply:
Actually `value_type` already exists (and is inherited from `TypeTensor<T>`)

Reply:
Agreed on the template.

In fact this wrapper only works on symmetric second order tensors. As you've probably already noticed, I am manually checking if the input second order tensor is symmetric. So although I may instantiate to `FactorizedRankTwoTensor`, it still really is a `FactorizedSymmetricRankTwoTensor` at its core, and so it may be a little misleading to name if `FactorizedRankTwoTensor`. I am with that though: we could name it `FactorizedRankTwoTensor`, and say it's the users responsibility to make sure that the tensor is symmetric.

Reply:
Another possibility is that I only instantiate a `FactorizedSymmetricRankTwoTensor` on `SymmetricRankTwoTensor`, and rely on the implicit conversion from `RankTwoTensor` to `SymmetricRankTwoTensor` at construction time. When the `RankTwoTensor` isn't symmetric, the implicit conversion fails.

I don't remember if your `SymmetricRankTwoTensor` stuff provides such implicit conversion though.

Reply:
It does not. I made those constructors `explicit`

```
  /// Copy constructor from TensorValue<T>
  explicit SymmetricRankTwoTensorTempl(const TensorValue<T> & a);

  /// Copy constructor from TypeTensor<T>
  explicit SymmetricRankTwoTensorTempl(const TypeTensor<T> & a);
```

Reply:
Yeah, it's dangerous if not being explicit.

Reply:
I'll do the explicit conversion in the constructor of this wrapper then. The check for symmetry is cheap compared to Eigen decomposition.

Reply:
> In fact this wrapper only works on symmetric second order tensors. As you've probably already noticed, I am manually checking if the input second order tensor is symmetric

If you use isSymmetric() for that check it works be a noop for the symmetric class and would still work for the regular class.

Reply:
It's templated on tensor type now.

Reply:
Wait a sec. I am getting compile errors in dbg saying `RankTwoTensorTempl` doesn't have a method `isSymmetric()`. Is it added in your Mandel notation PR?

Reply:
I also added the isSymmetric method.